src/zarr/core/array.py

```diff
@@ -945,7 +976,7 @@ def from_dict(
         ValueError
             If the dictionary data is invalid or incompatible with either Zarr format 2 or 3 array creation.
         """
-        metadata = parse_array_metadata(data)
+        metadata = parse_array_metadata(data)  # type: ignore[call-overload]
```

Comment:
Is the `type: ignore` needed due to the presence of `ArrayMetadata` in the signature for `__init__`?

Reply:
we need the overload because the `data` input to `from_dict` is typed as `dict[str, JSON]`. I could change that to the union of the two typeddict metadata types, but this will incompatibly override the base class implementation of `from_dict`, and so I will need another `# type: ignore` there. The only way to clean all of this up is to redo our base `Metadata` ABC, which I deemed out of scope for this PR